// @researchgate/react-history/cjs/BrowserHistory.js
"use strict";

exports.__esModule = true;

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _propTypes = require("prop-types");

var _propTypes2 = _interopRequireDefault(_propTypes);

var _history = require("history");

var _canUseDOM = require("./canUseDOM");

var _canUseDOM2 = _interopRequireDefault(_canUseDOM);

var _PropTypes = require("./PropTypes");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/**
 * Manages session history using the HTML5 history API including
 * pushState, replaceState, and the popstate event.
 */
var BrowserHistory = function (_React$Component) {
  _inherits(BrowserHistory, _React$Component);

  BrowserHistory.prototype.getChildContext = function getChildContext() {
    return { history: this.history };
  };

  function BrowserHistory(props) {
    _classCallCheck(this, BrowserHistory);

    var _this = _possibleConstructorReturn(this, _React$Component.call(this, props));

    if (_canUseDOM2.default) {
      var _this$props = _this.props,
          basename = _this$props.basename,
          forceRefresh = _this$props.forceRefresh,
          getUserConfirmation = _this$props.getUserConfirmation,
          keyLength = _this$props.keyLength;


      _this.history = (0, _history.createBrowserHistory)({
        basename: basename,
        forceRefresh: forceRefresh,
        getUserConfirmation: getUserConfirmation,
        keyLength: keyLength
      });

      // Do this here so we catch actions in cDM.
      _this.unlisten = _this.history.listen(function () {
        return _this.forceUpdate();
      });
    } else {
      _this.history = {};
    }
    return _this;
  }

  BrowserHistory.prototype.componentWillUnmount = function componentWillUnmount() {
    this.unlisten();
  };

  BrowserHistory.prototype.render = function render() {
    var children = this.props.children;


    return typeof children === "function" ? children(this.history) : _react2.default.Children.only(children);
  };

  return BrowserHistory;
}(_react2.default.Component);

BrowserHistory.propTypes = {
  basename: _propTypes2.default.string,
  forceRefresh: _propTypes2.default.bool,
  getUserConfirmation: _propTypes2.default.func,
  keyLength: _propTypes2.default.number,
  children: _propTypes2.default.oneOfType([_propTypes2.default.node, _propTypes2.default.func]).isRequired
};
BrowserHistory.childContextTypes = {
  history: _PropTypes.history.isRequired
};
exports.default = BrowserHistory;
