// @researchgate/react-history/cjs/Prompt.js
"use strict";

exports.__esModule = true;

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _propTypes = require("prop-types");

var _propTypes2 = _interopRequireDefault(_propTypes);

var _reactLifecyclesCompat = require("react-lifecycles-compat");

var _canUseDOM = require("./canUseDOM");

var _canUseDOM2 = _interopRequireDefault(_canUseDOM);

var _PropTypes = require("./PropTypes");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Prompt = function (_React$Component) {
  _inherits(Prompt, _React$Component);

  function Prompt(props, context) {
    _classCallCheck(this, Prompt);

    var _this = _possibleConstructorReturn(this, _React$Component.call(this, props, context));

    if (_canUseDOM2.default) {
      if (_this.props.when) {
        _this.enable(_this.props.message);
      }
    }
    return _this;
  }

  Prompt.prototype.enable = function enable(message) {
    if (this.unblock) this.unblock();
    this.unblock = this.context.history.block(message);
  };

  Prompt.prototype.disable = function disable() {
    if (this.unblock) {
      this.unblock();
      this.unblock = null;
    }
  };

  Prompt.prototype.getSnapshotBeforeUpdate = function getSnapshotBeforeUpdate(prevProps) {
    if (this.props.when) {
      if (!prevProps.when || prevProps.message !== this.props.message) this.enable(this.props.message);
    } else {
      this.disable();
    }
    return null;
  };

  Prompt.prototype.componentDidUpdate = function componentDidUpdate() {
    // we must define this lifecycle method as long as we're using the polyfill
  };

  Prompt.prototype.componentWillUnmount = function componentWillUnmount() {
    this.disable();
  };

  Prompt.prototype.render = function render() {
    return null;
  };

  return Prompt;
}(_react2.default.Component);

// Polyfill your component so the new lifecycles will work with older versions of React:


Prompt.propTypes = {
  when: _propTypes2.default.bool,
  message: _propTypes2.default.oneOfType([_propTypes2.default.func, _propTypes2.default.string]).isRequired
};
Prompt.contextTypes = {
  history: _PropTypes.history.isRequired
};
Prompt.defaultProps = {
  when: true
};
(0, _reactLifecyclesCompat.polyfill)(Prompt);

exports.default = Prompt;
