// @researchgate/react-history/cjs/PropTypes.js
"use strict";

exports.__esModule = true;
exports.history = exports.location = exports.action = undefined;

var _propTypes = require("prop-types");

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var action = exports.action = _propTypes2.default.oneOf(["PUSH", "REPLACE", "POP"]);

var location = exports.location = _propTypes2.default.shape({
  pathname: _propTypes2.default.string.isRequired,
  search: _propTypes2.default.string.isRequired,
  hash: _propTypes2.default.string.isRequired,
  state: _propTypes2.default.object,
  key: _propTypes2.default.string
});

var history = exports.history = _propTypes2.default.shape({
  action: action.isRequired,
  location: location.isRequired,
  push: _propTypes2.default.func.isRequired,
  replace: _propTypes2.default.func.isRequired,
  go: _propTypes2.default.func.isRequired,
  goBack: _propTypes2.default.func.isRequired,
  goForward: _propTypes2.default.func.isRequired,
  canGo: _propTypes2.default.func,
  block: _propTypes2.default.func.isRequired
});
